### PR TITLE
Hardening, 12th of October

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java [hunk 1]
```diff
@@ -362,82 +362,109 @@
...
+   */
+  @Deprecated
   public static final String CLASS_COMPLEXITY_KEY = "class_complexity";
-
   /**
    * Information about the cyclomatic complexity per class, calculated by divided the {@link #COMPLEXITY_IN_CLASSES} by the number of {@link #CLASSES}.
+   * @deprecated since 6.7
    */
+  @Deprecated
   public static final Metric<Double> CLASS_COMPLEXITY = new Metric.Builder(CLASS_COMPLEXITY_KEY, "Complexity / Class", Metric.ValueType.FLOAT)
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java]: MINOR Remove this use of "CLASS_COMPLEXITY_KEY"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java [hunk 2]
```diff
@@ -362,82 +362,109 @@
...
     .setDomain(DOMAIN_COMPLEXITY)
     .setDeleteHistoricalData(true)
+    .setHidden(true)
     .create();
 
+  /**
+   * @deprecated since 6.7
+   */
+  @Deprecated
   public static final String FUNCTION_COMPLEXITY_KEY = "function_complexity";
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java [hunk 3]
```diff
@@ -362,82 +362,109 @@
...
     .setDomain(DOMAIN_COMPLEXITY)
+    .setHidden(true)
     .create();
 
   /**
    * @since 3.6
+   * @deprecated since 6.7
    */
+  @Deprecated
   public static final String COMPLEXITY_IN_CLASSES_KEY = "complexity_in_classes";
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java [hunk 4]
```diff
@@ -451,22 +478,40 @@
       .setHidden(true)
       .create();
 
+  /**
+   * @deprecated since 6.7
+   */
+  @Deprecated
   public static final String FUNCTION_COMPLEXITY_DISTRIBUTION_KEY = "function_complexity_distribution";
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java [hunk 5]
```diff
@@ -362,82 +362,109 @@
...
+   */
+  @Deprecated
   public static final String FUNCTION_COMPLEXITY_KEY = "function_complexity";
-
   /**
    * Information about the cyclomatic complexity per function, calculated by divided the {@link #COMPLEXITY_IN_FUNCTIONS} by the number of {@link #FUNCTIONS}.
+   * @deprecated since 6.7
    */
+  @Deprecated
   public static final Metric<Double> FUNCTION_COMPLEXITY = new Metric.Builder(FUNCTION_COMPLEXITY_KEY, "Complexity / Function", Metric.ValueType.FLOAT)
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java]: MINOR Remove this use of "FUNCTION_COMPLEXITY_KEY"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java [hunk 6]
```diff
@@ -451,22 +478,40 @@
...
+  /**
+   * @deprecated since 6.7
+   */
+  @Deprecated
   public static final String FUNCTION_COMPLEXITY_DISTRIBUTION_KEY = "function_complexity_distribution";
+  /**
+   * @deprecated since 6.7
+   */
+  @Deprecated
   public static final Metric<String> FUNCTION_COMPLEXITY_DISTRIBUTION = new Metric.Builder(FUNCTION_COMPLEXITY_DISTRIBUTION_KEY, "Function Distribution / Complexity",
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java]: MINOR Remove this use of "FUNCTION_COMPLEXITY_DISTRIBUTION_KEY"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java [hunk 7]
```diff
@@ -362,82 +362,109 @@
...
    */
+  @Deprecated
   public static final String COMPLEXITY_IN_CLASSES_KEY = "complexity_in_classes";
 
   /**
    * @since 3.6
+   * @deprecated since 6.7
    */
+  @Deprecated
   public static final Metric<Integer> COMPLEXITY_IN_CLASSES = new Metric.Builder(COMPLEXITY_IN_CLASSES_KEY, "Complexity in Classes", Metric.ValueType.INT)
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java]: MINOR Remove this use of "COMPLEXITY_IN_CLASSES_KEY"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java [hunk 8]
```diff
@@ -362,82 +362,109 @@
...
     .setDomain(DOMAIN_COMPLEXITY)
+    .setHidden(true)
     .create();
 
   /**
    * @since 3.6
+   * @deprecated since 6.7
    */
+  @Deprecated
   public static final String COMPLEXITY_IN_FUNCTIONS_KEY = "complexity_in_functions";
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java [hunk 9]
```diff
@@ -362,82 +362,109 @@
     .setDomain(DOMAIN_COMPLEXITY)
     .create();
 
+  /**
+   * @deprecated since 6.7
+   */
+  @Deprecated
   public static final String FILE_COMPLEXITY_KEY = "file_complexity";
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java [hunk 10]
```diff
@@ -362,82 +362,109 @@
...
    */
+  @Deprecated
   public static final String COMPLEXITY_IN_FUNCTIONS_KEY = "complexity_in_functions";
-
   /**
    * @since 3.6
+   * @deprecated since 6.7
    */
+  @Deprecated
   public static final Metric<Integer> COMPLEXITY_IN_FUNCTIONS = new Metric.Builder(COMPLEXITY_IN_FUNCTIONS_KEY, "Complexity in Functions", Metric.ValueType.INT)
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java]: MINOR Remove this use of "COMPLEXITY_IN_FUNCTIONS_KEY"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java [hunk 11]
```diff
@@ -451,22 +478,40 @@
...
       .setQualitative(true)
       .setDomain(DOMAIN_COMPLEXITY)
+      .setHidden(true)
       .create();
 
+  /**
+   * @deprecated since 6.7
+   */
+  @Deprecated
   public static final String FILE_COMPLEXITY_DISTRIBUTION_KEY = "file_complexity_distribution";
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java [hunk 12]
```diff
@@ -362,82 +362,109 @@
...
+  @Deprecated
   public static final String FILE_COMPLEXITY_KEY = "file_complexity";
-
   /**
    * Information about the cyclomatic complexity per file, calculated by divided the {@link #COMPLEXITY} by the number of {@link #FILES}.
+   *
+   * @deprecated since 6.7
    */
+  @Deprecated
   public static final Metric<Double> FILE_COMPLEXITY = new Metric.Builder(FILE_COMPLEXITY_KEY, "Complexity / File", Metric.ValueType.FLOAT)
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java]: MINOR Remove this use of "FILE_COMPLEXITY_KEY"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java [hunk 13]
```diff
@@ -362,82 +362,109 @@
...
     .setDomain(DOMAIN_COMPLEXITY)
     .setDeleteHistoricalData(true)
+    .setHidden(true)
     .create();
 
+  /**
+   * @deprecated since 6.7
+   */
+  @Deprecated
   public static final String CLASS_COMPLEXITY_KEY = "class_complexity";
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java [hunk 14]
```diff
@@ -451,22 +478,40 @@
...
+  /**
+   * @deprecated since 6.7
+   */
+  @Deprecated
   public static final String FILE_COMPLEXITY_DISTRIBUTION_KEY = "file_complexity_distribution";
+  /**
+   * @deprecated since 6.7
+   */
+  @Deprecated
   public static final Metric<String> FILE_COMPLEXITY_DISTRIBUTION = new Metric.Builder(FILE_COMPLEXITY_DISTRIBUTION_KEY, "File Distribution / Complexity",
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/measures/CoreMetrics.java]: MINOR Remove this use of "FILE_COMPLEXITY_DISTRIBUTION_KEY"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 21 issues
* ![MINOR][MINOR] 7 minor
* ![INFO][INFO] 14 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-server/src/main/java/org/sonar/server/permission/ws/ProjectWsRef.java [hunk 15]
```diff
@@ -39,7 +39,7 @@
   private ProjectWsRef(@Nullable String uuid, @Nullable String key) {
     this.uuid = uuid;
     this.key = key;
-    checkRequest(this.uuid != null ^ this.key != null, "Project id or project key can be provided, not both.");
+    checkRequest(this.uuid != null ^ this.key != null, MSG_ID_OR_KEY_MUST_BE_PROVIDED);
```
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/permission/ws/ProjectWsRef.java]: This line makes line 54 of the same file obsolete. That line should be removed. We do the same check (but written differently) twice.

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/ShowAction.java [hunk 16]
```diff
@@ -83,11 +85,6 @@ public void handle(Request request, Response response) {
   }
 
   private static void checkOneOfIdOrNamePresent(@Nullable Long qGateId, @Nullable String qGateName) {
-    if (qGateId == null && qGateName == null) {
-      throw BadRequestException.create("Either one of 'id' or 'name' is required.");
-    } else if (qGateId != null && qGateName != null) {
-      throw BadRequestException.create("Only one of 'id' or 'name' must be provided.");
-    }
+    checkRequest(qGateId==null ^ qGateName==null, "Either '%s' or '%s' must be provided", PARAM_ID, PARAM_NAME);
```
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/ShowAction.java]: (tiny) I prefer spaces around `==`.

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchResponseFormat.java [hunk 17]
```diff
@@ -184,6 +184,7 @@ private void formatIssue(Issue.Builder issueBuilder, IssueDto dto, SearchRespons
       issueBuilder.setEffort(effortValue);
     }
     setNullable(dto.getLine(), issueBuilder::setLine);
+    setNullable(dto.getChecksum(), issueBuilder::setHash);
```
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchResponseFormat.java]: I tested with squid:S1220, it seems to produce null line numbers, but empty, non-null line hashes. I suggest to not output empty hash attributes in those cases.

File: sonar-core/src/main/resources/org/sonar/l10n/core.properties [hunk 18]
```diff
@@ -3,7 +3,6 @@
 # GENERIC WORDS, sorted alphabetically
 #
 #------------------------------------------------------------------------------
-
```
Comment (bartfastiel) [inline sonar-core/src/main/resources/org/sonar/l10n/core.properties]: A large number of empty lines was removed in this file. I just want to make sure, that it was on purpose.

File: server/sonar-db-dao/src/main/java/org/sonar/db/metric/MetricValidator.java [hunk 19]
```diff
@@ -60,4 +61,16 @@ static String validateDescription(@Nullable String description) {
...
   }
+
+  @CheckForNull
+  public static String checkMetricDomain(@Nullable String domain) {
+    if (domain == null) {
+      return null;
+    }
+
+    checkArgument(domain.length() <= MAX_DOMAIN_LENGTH, "Metric domain length (%s) is longer than the maximum authorized (%s). '%s' was provided.",
+      domain.length(), MAX_DOMAIN_LENGTH, domain);
```
Comment (bartfastiel) [inline server/sonar-db-dao/src/main/java/org/sonar/db/metric/MetricValidator.java]: This does not yet seem to work:
```
$ curl -s -u admin:admin -XPOST "localhost:9000/api/metrics/update?id=144&domain=aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
{"errors":[{"msg":"An error has occurred. Please contact your administrator"}]}
```

File: sonar-core/src/main/resources/org/sonar/l10n/core.properties (hunk 18 above)
Comment (teryk) [reply sonar-core/src/main/resources/org/sonar/l10n/core.properties]: good point, empty lines should not have been removed

File: sonar-core/src/main/resources/org/sonar/l10n/core.properties [hunk 20]
```diff
@@ -1728,8 +1728,8 @@ metric.new_major_violations.description=New Major issues
 metric.new_major_violations.name=New Major Issues
 metric.new_minor_violations.description=New Minor issues
 metric.new_minor_violations.name=New Minor Issues
-metric.new_lines.name=Lines of New Code
-metric.new_lines.description=Non commenting lines of new code
+metric.new_lines.name=New Lines
+metric.new_lines.description=New lines
```
[the remark this answers is not in the document]
Comment (teryk) [reply sonar-core/src/main/resources/org/sonar/l10n/core.properties]: of course there is, I don't know how I forgot to put it in the commit message: SONAR-9922

File: server/sonar-db-dao/src/main/java/org/sonar/db/metric/MetricValidator.java (hunk 19 above)
Comment (teryk) [reply server/sonar-db-dao/src/main/java/org/sonar/db/metric/MetricValidator.java]: Using exactly your curl command, I have the expected message…

Comment (bartfastiel) [reply server/sonar-db-dao/src/main/java/org/sonar/db/metric/MetricValidator.java]: I rebuilt the latest state of the branch, and now it also works fine for me. Sorry for the noise.

**Pull request merged**